### PR TITLE
Complete sanity functions for faster_rcnn and RetinaNet

### LINKED ISSUE #6: Retina Net
## 🚀 Feature

Our second detection model

The following needs to be done.

- [x] Support all torchvision backbones.
- [x] FPN Support for Resnet Backbones.
- [x] Support for negative samples etc.
- [x] Mixed Precision Training.
- [x] `train_step, val_step, fit ` APIs
- [ ] Sanity Fit API
- [x] IoU, MaP calculations while training and testing.
- [x] Lightning Trainer.

### LINKED ISSUE #7: Faster RCNN
## 🚀 Feature

Implement FRCNN completely.
Some code and tests are left

- [x] Support all torchvision backbones.
- [x] FPN Support for Resnet Backbones.
- [x] Support for negative samples etc.
- [x] Mixed Precision Training.
- [x] `train_step, val_step, fit ` APIs
- [ ] Sanity Fit API
- [x] IoU, MaP calculations while training and testing.
- [x] Lightning Trainer.

### PR DESCRIPTION
**Pull request by ramaneswaran**
Closes #7 
Closes #6

### CODE REVIEW

File: quickvision/models/detection/faster_rcnn/engine.py [hunk 1]
```diff
@@ -242,7 +242,29 @@ def train_sanity_fit(model, train_loader,
...
+    model = model.to(device)
+    model.train()
+
+    cnt = 0
+
+    for batch_idx, (input, targets) in enumerate(train_loader):
+        
+        images = list(image.to(device) for image in outputs)
+        targets = [{k: v.to(device) for k, v in t.items()} for t in targets]
+
```
Comment (oke-aditya) [inline quickvision/models/detection/faster_rcnn/engine.py:255]:
```
 if scaler is not None:
            with amp.autocast():
                loss_dict = model(images, targets)
```

Can we add `amp` too for `fp16` check ?

File: quickvision/models/detection/faster_rcnn/engine.py [hunk 2]
```diff
@@ -242,7 +242,29 @@ def train_sanity_fit(model, train_loader,
...
+        loss_dict = model(images, targets)
+
+        cnt += 1
+
```
Comment (oke-aditya) [inline quickvision/models/detection/faster_rcnn/engine.py:259]:
```
if last_batch or batch_idx % log_interval == 0:
                print(f"Validation Sanity check passed for batch till {batch_idx} batches")
```

File: quickvision/models/detection/faster_rcnn/engine.py [hunk 3]
```diff
@@ -242,7 +242,29 @@ def train_sanity_fit(model, train_loader,
...
+        if num_batches is not None:
+            if cnt >= num_batches:
+                print(f"Done till {num_batches} train batches")
```
Comment (oke-aditya) [inline quickvision/models/detection/faster_rcnn/engine.py:262]:
```
print("All specicied batches done.")
val_sanity_end = time.time()
print(f"Validation Sanity check passed in time {val_sanity_end - val_sanity_start} !!")
```

File: quickvision/models/detection/faster_rcnn/engine.py [hunk 4]
```diff
@@ -262,8 +284,28 @@ def val_sanity_fit(model, val_loader,
...
+            cnt += 1
+
+            if num_batches is not None:
+                if cnt >= num_batches:
+                    print(f"Done till {num_batches} validation batches")
+                    break
+
+            
+    
+    return True
```
Comment (oke-aditya) [inline quickvision/models/detection/faster_rcnn/engine.py:309]: Same changes as above.
Not calculating metrics, or `AverageMeter` is fine.

Make CI pass, and add tests. Once we do that, the same code can be duplicated for retinanet.

**Review by oke-aditya — CHANGES_REQUESTED** [review]
Need small changes @ramaneswaran . The overall logic is correct and code is completely fine 😄

File: quickvision/models/detection/faster_rcnn/engine.py (hunk 1 above)
Comment (ramaneswaran) [reply quickvision/models/detection/faster_rcnn/engine.py:255]: Noted

File: quickvision/models/detection/faster_rcnn/engine.py (hunk 2 above)
Comment (ramaneswaran) [reply quickvision/models/detection/faster_rcnn/engine.py:259]: All right last batch, got it

File: quickvision/models/detection/faster_rcnn/engine.py (hunk 3 above)
Comment (ramaneswaran) [reply quickvision/models/detection/faster_rcnn/engine.py:262]: But function description mentions time and metrics are not tracked. 
I will add the time calculation too

File: quickvision/models/detection/faster_rcnn/engine.py (hunk 4 above)
Comment (ramaneswaran) [reply quickvision/models/detection/faster_rcnn/engine.py:309]: Noted

File: quickvision/models/detection/faster_rcnn/engine.py (hunk 3 above)
Comment (oke-aditya) [reply quickvision/models/detection/faster_rcnn/engine.py:262]: Time is fine, it just shows a rough estimate of code. Sometimes it is too long to do a sanity fit too.
The `log_interval` would then keep showing status to user.

File: quickvision/models/detection/faster_rcnn/lightning_trainer.py [hunk 5]
```diff
@@ -4,7 +4,11 @@
 from quickvision.models.components import create_torchvision_backbone
```
Comment (oke-aditya) [inline quickvision/models/detection/faster_rcnn/lightning_trainer.py:4]: nit: Can you undo for this file. It is not needed to change this I guess.

File: quickvision/models/detection/faster_rcnn/model_factory.py [hunk 6]
```diff
@@ -1,13 +1,18 @@
-
 import torch.nn as nn
-from torchvision.models.detection.faster_rcnn import fasterrcnn_resnet50_fpn, FasterRCNN, FastRCNNPredictor
+from torchvision.models.detection.faster_rcnn import (
```
Comment (oke-aditya) [inline quickvision/models/detection/faster_rcnn/model_factory.py:2]: nit: You needn't modify formatting for this file as well.

File: tests/test_frcnn.py [hunk 7]
```diff
@@ -168,21 +171,51 @@ def test_fit_cuda(self):
...
 
     @unittest.skipIf(not torch.cuda.is_available(), "CUDA unavailable")
     def test_train_sanity_fit_cuda(self):
-        pass
+        for bbone in fpn_supported_models:
+            backbone = faster_rcnn.create_fastercnn_backbone(backbone=bbone, pretrained=None)
+            self.assertTrue(isinstance(backbone, nn.Module))
+            frcnn_model = faster_rcnn.create_vision_fastercnn(num_classes=3, backbone=backbone)
+            self.assertTrue(isinstance(frcnn_model, nn.Module))
+            result = faster_rcnn.train_sanity_fit(frcnn_model, train_loader, "gpu", num_batches=10)
```
Comment (oke-aditya) [inline tests/test_frcnn.py:189]: `cuda` not `gpu`

File: tests/test_frcnn.py [hunk 8]
```diff
@@ -168,21 +171,51 @@ def test_fit_cuda(self):
...
 
     @unittest.skipIf(not torch.cuda.is_available(), "CUDA unavailable")
     def test_sanity_fit_cuda(self):
-        pass
+        for bbone in fpn_supported_models:
+            backbone = faster_rcnn.create_fastercnn_backbone(backbone=bbone, pretrained=None)
+            self.assertTrue(isinstance(backbone, nn.Module))
+            frcnn_model = faster_rcnn.create_vision_fastercnn(num_classes=3, backbone=backbone)
+            self.assertTrue(isinstance(frcnn_model, nn.Module))
+            result = faster_rcnn.sanity_fit(frcnn_model, train_loader, "gpu", num_batches=10)
```
Comment (oke-aditya) [inline tests/test_frcnn.py:217]: `cuda` not `gpu`

File: quickvision/models/detection/faster_rcnn/engine.py [hunk 9]
```diff
@@ -262,8 +295,37 @@ def val_sanity_fit(model, val_loader,
...
+            last_batch = batch_idx == last_idx
+            images = list(image.to(device) for k,v in t.items() for t in targets)
+            targets = [{k: v.to(device) for k, v in t.items()} for t in targets]
+
+            out = model(images)
 
+            cnt += 1
+
+
+            if last_batch or batch_idx%log_interval == 0:
```
Comment (oke-aditya) [inline quickvision/models/detection/faster_rcnn/engine.py:314]: nit: A bracket and space will be nice.

**Review by oke-aditya — CHANGES_REQUESTED** [review]
Most things look very good to me.  A few bugs in CUDA tests.
There are small nits do let me know.

I don't use Black formatting and don't prfer it as well. You can undo those formatting changes too.

**Comment by oke-aditya** [comment]
There is CI failure due to a small `}` can you fix that as well ?
Can't merge without the required CI changes too.

**Comment by oke-aditya** [comment]
I will complete this up for FRCNN and RetinaNet 👍

**Review by oke-aditya — APPROVED** [review]
LGTM !
Just Detr Left which I will complete.
Then we are almost there for release !

**Pull request merged**